### DISCUSSION TITLE
Is changing dependency constraints of already previously published versions of a package a bad practice or not?

### DISCUSSION THREAD

**Discussion by hostep** [composer/composer#10063] [General]
Hi guys

I've now seen this behavior twice by two separate companies who host their own private composer repository (unsure if they use private packagist, but I guess they don't).

So, here's what I see:
- At a certain time, they publish a new version (1.4.3) of some package (`vendor/example1`)
- That package has a dependency on another package (`vendor/example2`) and a certain constraint for that other dependency, let's say `>=6.0.3`
- Then for some reason, a couple of weeks/months later, they change that version constraint to `>=7.0.1` in their `https://repo.vendor.com/packages.json` file still for the same 1.4.3 version of `vendor/example1`

We notice this problem only with composer version 2, when we try to update the `composer.lock` file using `composer update --lock` command (because we sometimes have to solve git merge conflicts and let composer recalculate the hash). Where we then get an error like this:

```
$ composer update --lock
Your requirements could not be resolved to an installable set of packages.
  Problem 1
    - Root composer.json requires vendor/example1 == 1.4.3.0 -> satisfiable by vendor/example1[1.4.3].
    - vendor/example1 1.4.3 requires vendor/example2 >=7.0.1 -> found vendor/example2[7.0.1] but these were not loaded, likely because it conflicts with another require.
```

So the question: is it considered bad practice to change dependency constraints in a certain version of a package, without also publishing a new version of that package?
To me it feels very wrong.

Thanks!

**Comment by Seldaek** [comment]
That definitely is quite a bad practice IMO. Tags should be set in stone, and re-tagging, whether it's to update dependencies, or the code of the tag, is a terrible idea. Shipping a patch (1.4.4) release that requires a higher version sounds okish, although usually I would reserve that for a minor (1.5.0) release.

**Reply by hostep** [reply]
Thanks for the reply Jordi

Actually, they don't re-tag as far as I can see, they just rewrite the `packages.json` file on their repo with updated dependency information.

Luckily one of both companies where I reported the problem today, already confirmed and fixed the issue from their side.
I'm going to try to convince the other company again where I initially reported the problem to a month ago, they don't seem to understand the impact of the problem yet.

**Reply by hostep** [reply]
Huh, I just noticed today that they do actually re-tag versions. The same version downloaded months ago versus the same version downloaded today contains difference in their copyright headers in all their php files (and here and there version requirement changes in the `composer.json` files).

This is so stupid, the guys that maintain https://composer.amasty.com/ really need to stop doing this...

**Comment by willrowe** [comment]
@Seldaek I ran into this situation today and I'm wondering if there's any good way to handle it. The [rcrowe/twigbridge](https://packagist.org/packages/rcrowe/twigbridge) package prior to v0.14.3 required `twig/twig: ~3.0`. However, in v3.9 of `twig/twig` a change was made that broke `rcrowe/twigbridge`, because they were using an internal function. This was [fixed](https://github.com/rcrowe/TwigBridge/pull/441), but there was no way for me to know that there would be an issue when I updated `twig/twig` to v3.14.0 in order to patch a security vulnerability. Making the updated then caused a crash on production.

All of the versions before v0.14.3 that `rcrowe/twigbridge` released that included the `~3.0` constraint are not actually compatible with `twig/twig` v3.9+, but Composer will still install the latest version. Is there any way to go back and mark those older versions as being incompatible with v3.9+?

**Reply by stof** [reply]
except by rewriting the history of your git repo to change the tags, no. but rewriting the history of tags will cause lots of issues (git does not help you at all when you try to do that, as it treats tags as immutable so does not check them for rewrites)

**Reply by Seldaek** [reply]
Yeah this is an unfortunate situation that sometimes happens and the only good solution is to move forward and upgrade all deps that need upgrading together. Fixing the past there is not really a realistic option.
